icicle/include/api/babybear.h

```diff
@@ -32,47 +32,49 @@ extern "C" cudaError_t babybear_create_poseidon2_cuda(
   device_context::DeviceContext& ctx
 );
 
-extern "C" cudaError_t babybear_init_poseidon2_cuda(
+extern "C" cudaError_t babybear_poseidon2_load_cuda(
   poseidon2::Poseidon2<babybear::scalar_t>** poseidon,
   unsigned int width,
   poseidon2::MdsType mds_type,
   poseidon2::DiffusionStrategy diffusion,
   device_context::DeviceContext& ctx
 );
 
-extern "C" cudaError_t babybear_poseidon2_permute_many_cuda(
+extern "C" cudaError_t babybear_poseidon2_absorb_many_cuda(
+  const poseidon2::Poseidon2<babybear::scalar_t>* poseidon,
+  const babybear::scalar_t* inputs,
+  babybear::scalar_t* states,
```

Comment:
what if states was internal to poseidon? I mean the first param. Or do you prefer to decouple them? why?

Reply:
the idea here is to give a flexible API for user to do any kind of memory operations. If you look into our tree builder for example, it features a lot of memory handling tricks, like using just 2 vectors to build an entire tree. Also, I don't really like adding states inside the class as it will require managing the release of that memory through ffi